A refined CPU set iterator

arthurpaulino opened the pull request:
The current way to iterate over a CPU set has two problems:
1. It spends too many cycles on bits associted with inactive CPUs
2. It entails an ad-hoc solution for the round-robin iteration in the CPU selection algorithm, involving a chain of two filtered iterators and using -1 to mark the absence of a last chosen CPU, resulting on a non-idiomatic implementation

This patch implements the `CpuCycler`, an iterator that can wrap around the CPU set while skipping chunks of inactive CPUs from the same `InnerPart`.

This new iterator mitigates the first problem while also providing an API that naturally solves the second problem.

lrh2000 reviewed (commented):
This sounds like a good idea. Thanks for the work.

arthurpaulino commented on ostd/src/cpu/set.rs:
```diff
@@ -303,3 +382,84 @@ mod test {
+        set
+    }
+
+    #[test]
```
I thought `ktest` was only for tests that required the initiation of some specific entities in the kernel, which is not the case for the specific tests I added.

There are more instances of `#[test]` in the codebase.

tatetian replied:
The `#test` test cases are run with `cargo test`, while the `#ktest` ones are done with `cargo ktest`. The former ones are run in a Linux process, the latter ones are run an OSTD-managed VM. In practice, this means that normal Cargo-managed crates should be tested with `cargo test`, yet OSDK-managed crates should `cargo ktest`.

See the definitions of `NON_OSDK_CRATES` and `OSDK_CRATES` in Makefile and how they are used by `make test` and `make ktest`:

https://github.com/asterinas/asterinas/blob/56d78ab64965194ad03cb6cde07fd5d98b17129d/Makefile#L303-L318

If you see any use of `#[test]` in OSTD-based crates or the OSTD crate itself, then it must be a bug. Please report or fix them.

arthurpaulino commented on kernel/src/sched/sched_class/mod.rs:
```diff
@@ -285,23 +285,15 @@ impl ClassScheduler {
-            None => -1,
-        };
-        // Simulate a round-robin selection starting from the last chosen CPU.
+
```
I'm not sure I understand the rules for blank lines. Please let me know if there are guidelines.

tatetian commented:
This PR is well intended. The new `iter_after` API does improve the readability of the scheduler code. But the new API is introduced at the expense of increasing the size of OSTD, which is the memory-safety TCB.

Note that OSTD already includes the `bitvec` crate as part of its TCB:

https://github.com/asterinas/asterinas/blob/56d78ab64965194ad03cb6cde07fd5d98b17129d/ostd/libs/id-alloc/src/lib.rs#L13-L16

and

https://github.com/asterinas/asterinas/blob/56d78ab64965194ad03cb6cde07fd5d98b17129d/ostd/src/cpu/local/dyn_cpu_local.rs#L124-L127

So do downstream crates that use OSTD.

https://github.com/asterinas/asterinas/blob/56d78ab64965194ad03cb6cde07fd5d98b17129d/kernel/src/fs/exfat/bitmap.rs#L26-L29

and

https://github.com/asterinas/asterinas/blob/56d78ab64965194ad03cb6cde07fd5d98b17129d/kernel/comps/network/src/dma_pool.rs#L139-L146

**So what if we could utilize the existing ability offered by `bitvec`? This way, we can enjoy the convenience of more user-friendly APIs without bloating the memory-safety TCB.**

This PR, in essence, implements a bit iterator with some specialized functionality (iterating from a specified offset). In theory, this PR's optimized iterator implementation should have already been present in the `bitvec` crate. We should be able to implement our hand-crafted `iter_after` method by combining `bitvec`'s [`iter_ones`](https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html#method.iter_ones) and [`get`](https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html#method.get) APIs and doing so should be as efficient as our hand-crafted `iter_after`.

So I propose exposing the underlying slice of a `CpuSet`. 

```rust
impl CpuSet {
    /// Returns a reference to the underlying slice that backs the `CpuSet`.
    ///
    /// The returned `u64` slice is a bitmap with 
    /// bit 1 indicates the CPU at the offset is present in the CPU set
    /// and bit 0 means that the CPU is not present.
    /// The bits beyond the first `ostd::cpu::num_cpus` bits are all zeros and should be ignored.
    pub fn as_slice(&self) -> &[u64] {
        self.bits.as_slice()
    }
}
```

In the scheduler subsystem of the kernel crate, you can implement an extension trait for `CpuSet` that provides the `iter_after` method:

```rust
trait CpuSetExt {
    fn iter_after(&self, after: CpuId) -> impl Iterator<Item = CpuId> + '_ {
        let bitmap = self.as_slice();
        todo!("convert bitmap to `BitSlice`")
    }
}
```

The `bitvec` crate supports converting an existing slice into a `BitSlice`, which provides the `get` and `iter_one` APIs that I mentioned before.

lrh2000 commented:
The `CpuSet::iter` method must reside in OSTD because some users are in OSTD. Therefore, it deserves a better implementation. Meanwhile, `iter` and `iter_after` can be implemented similarly. It may not be a bad idea to put `iter_after` in OSTD, though I agree that it is indeed questionable whether OSTD should conceptablly have that functionality.

As a side note, I was surprised to find that [`bitvec` is nearly unmaintained now](https://github.com/ferrilab/bitvec) (although [it's still quite popular](https://lib.rs/crates/bitvec)). It also has some complicated features that we may not need (e.g., atomic support?). So I'm also unsure about how we should live with it in the future.

arthurpaulino commented:
> `fn iter_in(&self, range: Range<CpuId>)`

It's not clear to me how that helps when building a cycling API.

This patch touches constraints related to TCB size, placement of implementations and dependencies used. For the sake of productivity, I'd rather defer it to someone with more context (@lrh2000 feel free to take over if you'd like).

The fundamental problems are clear:
> 1. It spends too many cycles on bits associted with inactive CPUs
> 2. It entails an ad-hoc solution for the round-robin iteration in the CPU selection algorithm, involving a chain of two filtered iterators and using -1 to mark the absence of a last chosen CPU, resulting on a non-idiomatic implementation

I'd be happy with any solution that addresses those issues.

lrh2000 commented:
> It's not clear to me how that helps when building a cycling API.

Cycling is something like:
```rust
impl CpuSetExt for CpuSet {
    fn cycle_after(&self, after: CpuId) -> impl Iterator<Item = CpuId> {
        cpu_set.iter_in(after + 1..).chain(cpu_set.iter_in(0..=after))
   }
}
```

I believe it's what @tatetian means by saying:
>> We should be able to implement our hand-crafted `iter_after` method by combining `bitvec`'s [`iter_ones`](https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html#method.iter_ones) and [`get`](https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html#method.get) APIs and doing so should be as efficient as our hand-crafted `iter_after`.

But `Range<CpuId>` is indeed awkward to be used because `after + 1` simply does not compile and we cannot make it compile (`CpuId` must correspond to a real CPU: if `cpu_id` is the last CPU, `cpu_id + 1` should fail). To resolve the problem, we can adjust the signature of `CpuSet::iter_in` to accept a `Range<usize>`, but then the API becomes (slightly) odd.

> @lrh2000 feel free to take over if you'd like

Let's keep this PR open and wait for @tatetian's further comments first.

tatetian commented:
@lrh2000 has summarized my points well.

The `iter_after` operation is conceptually not a fundamental operation as it can be implemented by combining more basic iterators such as `iter_in`. One can implement `iter_after` easily and efficiently with `iter_in`, but not the other way around.

As an alternative, the `iter_after` operation can be implemented efficiently by an OSTD-based kernel using on the underlying slice exposed through `CpuSet::as_slice`.

Either way, we should only add essential/fundamental APIs to OSTD types and make these APIs flexible, composable, and powerful to implement conceptually more complex operations.

> The `CpuSet::iter` method must reside in OSTD because some users are in OSTD. Therefore, it deserves a better implementation. Meanwhile, `iter` and `iter_after` can be implemented similarly.

This is a good point. A more efficient iterator should be shipped with the OSTD.

> But `Range<CpuId>` is indeed awkward to be used because `after + 1` simply does not compile and we cannot make it compile (`CpuId` must correspond to a real CPU: if `cpu_id` is the last CPU, `cpu_id + 1` should fail). To resolve the problem, we can adjust the signature of `CpuSet::iter_in` to accept a `Range<usize>`, but then the API becomes (slightly) odd.

We can the signature of the `iter_in` method to this.

```rust
impl CpuSet {
    fn iter_in(&self, range: I) -> impl Iterator<Item = CpuId> 
    where
        I: SliceIndex<CpuSet>
    {
        todo("implement this via `BitSlice::get` and `BitSlice::iter_ones`")
    }
}

/// A trait that unifies all types that index a portion of `CpuSet`.
pub trait CpuSetSliceIndex {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb>;
}

impl CpuSetSliceIndex for RangeTo<CpuId> {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb> { todo!() }
}

impl CpuSetSliceIndex for RangeFrom<CpuId> {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb> { todo!() }
}

impl CpuSetSliceIndex for Range<CpuId> {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb> { todo!() }
}

impl CpuSetSliceIndex for RangeFull<CpuId> {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb> { todo!() }
}

impl CpuSetSliceIndex for RangeToInclusive<CpuId> {
    fn to_bit_slice_index(self) -> BitSliceIndex<'a, u64, Lsb> { todo!() }
}
```

This way, the scheduler can implement `cycle_after` this way.

```rust
impl CpuSetExt for CpuSet {
    fn cycle_after(&self, after: CpuId) -> impl Iterator<Item = CpuId> {
        cpu_set.iter_in((Bound::Excluded(after), Bound::Unbounded))
            .chain(cpu_set.iter_in(0..=after))
   }
}
```

tatetian commented on ostd/src/cpu/set.rs:
```diff
@@ -112,16 +120,25 @@ impl CpuSet {
+
+        bit_slice[bit_slice_index]
+            .iter_ones()
+            .map(|bit_pos| CpuId(bit_pos as u32))
```
I don't know the exact reason of the failure that you asked. But this implementation is buggy.

```rust
impl CpuSet {
    pub fn new_full() -> Self {
        let mut ret = Self::with_capacity_val(num_cpus(), !0);
        ret.clear_nonexistent_cpu_bits();
        ret
    }
}
```

Note that this constructor fills the integer array with all bits of 1, even the ones that are out of the valid CPU ID range. This means that `CpuSet::iter` may return invalid `CpuId`s. For example, assuming that you have a CPU of 4 cores, what value are expecting for `self.bits`? `0x0000_0000_0000_000F` or `0xFFFF_FFFF_FFFF_FFFF`? By calling `iter_onces` directly, you have assumed that it is the former. But in reality, `CpuSet::new_full` has initialized the value to the former one. This is not a problem in the past because the extra bits are never used. But now this assumption is broken by the use of `bitvec`.

This bug highlights the need for adding some unit tests, where `CpuSet::iter` is used on an instance created by `CpuSet::new_full`. Bit operations and iterators are notoriously difficult to get it right. Take extra care when writing such code.

lrh2000 replied:
> But in reality, `CpuSet::new_full` has initialized the value to the former one.

But...
https://github.com/asterinas/asterinas/blob/48ea2248953f86d1ece3015f871879aa7e3a5bc2/ostd/src/cpu/set.rs#L153-L159

Have I misunderstood the code? The unused bits are cleared here.

The issue seems to be solved? Not really sure what's the root cause..

arthurpaulino commented on ostd/src/cpu/set.rs:
```diff
@@ -108,20 +112,34 @@ impl CpuSet {
+            .as_slice()
+            .view_bits::<Lsb0>()
+            .iter_ones()
+            .filter_map(move |bit_pos| {
```
Good catch! Patched.

lrh2000 commented on ostd/src/cpu/set.rs:
```diff
@@ -108,20 +112,49 @@ impl CpuSet {
+            .iter()
+            .enumerate()
+            .filter_map(move |(part_idx, part)| {
+                if *part != 0 && part_idx >= part_idx_min && part_idx <= part_idx_max {
```
This seems still unnecessarily complex. I tried this. It seems to work:
```rust
    pub fn iter_in<I: CpuSetSlicer>(&self, slicer: I) -> impl Iterator<Item = CpuId> + '_ {
        let (start, end) = slicer.to_range_bounds();

        // Compute the minimum and maximum part indices.
        let part_idx_min = (start / BITS_PER_PART).min(self.bits.len());
        let part_idx_max = end.div_ceil(BITS_PER_PART).min(self.bits.len());

        let offset = part_idx_min * BITS_PER_PART;
        self.bits[part_idx_min..part_idx_max]
            .view_bits::<Lsb0>()
            .iter_ones()
            .filter_map(move |bit_part_idx| {
                let bit_idx = offset + bit_part_idx;
                if bit_idx >= start && bit_idx < end {
                    Some(CpuId(bit_idx as u32))
                } else {
                    None
                }
            })
    }
```

lrh2000 commented on ostd/src/cpu/set.rs:
```diff
@@ -245,30 +337,207 @@ mod test {
+    fn range_from_larger_than_max_cpu() {
+        // Only has CPUs 0-2
+        let set = make_cpu_set(&[0, 1, 2]);
+        // RangeFrom with start beyond any existing CPU
```
```suggestion
        // `RangeFrom` with start beyond any existing CPU
```

arthurpaulino commented:
@boterinas rerun

tatetian reviewed (commented):
The code looks mostly great. Thanks for your effort!

arthurpaulino commented on ostd/src/cpu/set.rs:
```diff
@@ -141,6 +163,63 @@ impl CpuSet {
+
+impl CpuSetSlicer for RangeToInclusive<CpuId> {
+    fn to_range_bounds(self) -> (usize, usize) {
+        (0, self.end.as_usize() + 1)
```
I think doing saturating addition is a healthy measure in these cases. Thanks for pointing that out! I'm going to change it when I have a chance.

lrh2000 replied:
> I think doing saturating addition is a healthy measure in these cases.

I think it's confusing because it makes the reader think the case could happen. However, it should be part of the `CpuId`'s type invariant.

We should answer the following questions when considering this:
 - Is this worth checking in the release build?
   - My answer is no.
 - If we decide to check it, what should the correct behavior be?
   - My answer is to panic (**instead of saturating**, especially in debug builds) because it's a bug.

Therefore, it does seem that debug-build assertions are most suitable.

To clarify this, perhaps we should have a comment:
```rust
        // The type invariant of `CpuId` ensures that the ID is always a resonable value.
        // So adding one to it will not cause an overflow.
        (0, self.end.as_usize() + 1)
```

arthurpaulino replied:
What about 32-bit architectures?

lrh2000 replied:
> The type invariant of `CpuId` ensures that the ID is **always a reasonable value**.

By "reasonable value," I mean the value should be small. It should not be something like `usize::MAX` or `usize::MAX - 1`, nor should it be something like `u32::MAX` or `u32::MAX - 1`.

For example, we can assume that there are at most 10⁸ CPUs. If there are more, we will ignore them. I believe this is a reasonable assumption. The architectures being 32-bit or 64-bit is entirely unrelated.

I can provide many justifications for this. The two most important are:
 - Each CPU has to have some CPU-local resources, which takes some memory space (and certainly need more than one byte). So if we're in 32-bit architectures and there are `u32::MAX` CPUs, we cannot even allocate enough memory to store the CPU-local resources!
 - The Rust language [assumes that all pointer offsets cannot exceed `isize::MAX` (note it's **not** `usize::MAX`)](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.offset). So if there are more than `isize::MAX` CPUs, we will no longer be allowed to use the CPU ID as the pointer offset or array indices. Instead of fighting with the Rust compiler, which can easily cause undefined behavior, we should just forbid this from happening.

tatetian commented on ostd/src/cpu/set.rs:
```diff
@@ -108,20 +112,38 @@ impl CpuSet {
+        let part_idx_max = end.div_ceil(BITS_PER_PART).min(self.bits.len());
+
+        let offset = part_idx_min * BITS_PER_PART;
+        self.bits[part_idx_min..part_idx_max]
```
Thanks guys. I now have a better understanding what is going on here.

> We need `min` on `start` too otherwise `.iter_in(CpuId(u32::MAX)..)` would panic

This is not true. **We should NOT create `CpuId` instances with invalid CPU ID numbers, which breaks the invariant of `CpuId`**, as documented by its Rust doc.

```rust
/// The ID of a CPU in the system.
///
/// If converting from/to an integer, the integer must start from 0 and be less
/// than the number of CPUs.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CpuId(u32);
```

The expression of `CpuId(u32::MAX)` is not even compilable outside the `cpu` module of OSTD. Unfortunately, this PR uses the internal, unchecked construction expression `CpuId(an_integer_value)` quite freely in `set.rs`. The test cases include expressions such as `CpuId(100)`. **These test cases should be rewritten to avoid creating arbitrary CPU IDs that are invalid on the current platform.** Since the total number of CPU cores is determined at run time, these test cases should not be written to rely on such a parameter.

To avoid making the same mistake in the future, I suggest moving `CpuId` to its own sub-module so that other sub-modules under the parent `cpu` module cannot instantiate `CpuId` directly. I will submit a PR.

lrh2000 replied:
> To avoid making the same mistake in the future, I suggest moving `CpuId` to its own sub-module so that other sub-modules under the parent `cpu` module cannot instantiate `CpuId` directly.

I fully agree.

> But my goal with the instantiation of `CpuId` with dummy raw IDs and `CpuSet` with arbitrary CPUs was to validate the pure logic of `iter_in` in a wider range of scenarios

But I also agree that creating invalid CPU IDs for this purpose (unit tests) is fine. I cannot think of a better way to do it. Do we really care about the soundness properties of unit testing environments (as long as no memory safety problems are triggered)? It seems quite unnecessary to me...

arthurpaulino replied:
> This is not true. We should NOT create CpuId instances with invalid CPU ID numbers, which breaks the invariant of CpuId, as documented by its Rust doc.

I only used `u32::MAX` for the sake of expediency. It's still possible to make the code without `min` on `start` panic with only valid CPUs. For example, when `start` comes from a `Bound(Excluded(cpu_id))` and `cpu_id` is the very last CPU in the set. `start` will be out of bounds by exactly 1, because "excluded" on start adds 1 to the index.

lrh2000 replied:
> I can create an auxiliary `dummy_cpu` function for testing purposes that unsafely calls `new_unchecked` so the current test coverage can remain the same

I am also 100% fine with this. However, #2514 proposes adding a generic parameter (`IdSet<T>`, see the third point in the PR description for more details), which seems overly complex.

@tatetian What is your opinion on the other "bugs" relating to soundness in the unit tests? For example, when testing the page table, we write:
https://github.com/asterinas/asterinas/blob/a7e45201e739ab20d43418fb6dd3839e3c326b1d/ostd/src/mm/page_table/test.rs#L41-L52
I don't think it's worth "fixing". It's fine to write unsafe code more freely in unit tests as long as it doesn't actually trigger UBs.

arthurpaulino replied:
> I am also 100% fine with this. However, #2514 proposes adding a generic parameter (`IdSet<T>`, see the third point in the PR description for more details), which seems overly complex.

Right, the third commit of #2514 hadn't been pushed when I wrote the comment above.

I don't have enough context to argue against (or in favor of) added complexity in the project. But I did point out one issue I could see there.

tatetian replied:
> I am also 100% fine with this. However, #2514 proposes adding a generic parameter (`IdSet<T>`, see the third point in the PR description for more details), which seems overly complex.

I think we can agree that breaking some type variants in unit tests is tolerable but unsatisfying, right? If we have the chance to make the code more elegant, why not?

Let's analyze the pros and cons of introducing `IdSet<I>` and `Id`.
* Pros:
    1. **The code becomes more elegant** because it promotes **separation of concerns**: all set operations are implemented by `IdSet<I>` and `Id`, whereas the CPU-related operations are implemented by `CpuSet` and `CpuId`.
    2. **The code becomes more flexible** as one may reuse `IdSet<I>` to implement more container type for IDs. For example, when we have NUMA support, we may want to have a container for NUMA node IDs. Obviously, we do not want to duplicate code `CpuSet`. A container type for IDs is quite a common concept.
    3. **The code becomes more testable**, which is our original motivation.
 * Cons:
    1. **The code becomes more complex**. But the complexity growth is quite acceptable. The implementation of `IdSet` is almost identical to the old `CpuSet`; the only difference is the use of a generic type parameter `I`.
   
I think the biggest obstacle of introducing the `Id` and `IdSet<I>` abstractions is that someone needs to take the effort of actually realizing the idea. And I have done just that.

tatetian commented on ostd/src/util/id_set.rs:
```diff
@@ -241,30 +243,97 @@ impl<I: Id> IdSet<I> {
+        // Compute the inbound start and end indices.
+        // Out-of-bounds indices are allowed. They will be ignored.
+        let inbound_start = start.min(self.bits.len() * BITS_PER_PART);
+        let inbound_end = end.min(self.bits.len() * BITS_PER_PART);
```
> That's _almost_ correct, except that we don't have access to an `I::cardinality` when implementing `IdSetSlicer for RangeFull`

Haha. Ok. This is a surprise. How about adding a generic parameter `I` to `IdSetSlicer<I>`? This makes sense as `IdSetSlicer` serves `IdSet<I>`, which has the generic parameter `I`.

tatetian commented on ostd/src/util/id_set.rs:
```diff
@@ -241,30 +243,97 @@ impl<I: Id> IdSet<I> {
+
+/// Returns an [`usize`] from the [`Id`]'s corresponding [`u32`].
+fn id_as_usize<I: Id>(id: I) -> usize {
+    Into::<u32>::into(id) as usize
```
How about add `Into<usize>` as a super trait of `Id` in addition to `Into<u32>`?

lrh2000 replied:
`From`/`Into` is known to behave poorly for type inference. So unfortunately, introducing multiple `Into<{num}>` traits on the same type is a bad idea. For example, even the simple code you suggested will fail to compile ([Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=56318adc46597a5e07a71ae633500a9e)):
```rust
struct A;
impl From<A> for u32 {
    fn from(value: A) -> Self {
        1
    }
}
impl From<A> for usize {
    fn from(value: A) -> Self {
        1
    }
}

fn test(a: A) -> usize {
    a.into() + 1
}
```
```
error[E0283]: type annotations needed
  --> src/lib.rs:14:7
   |
14 |     a.into() + 1
   |       ^^^^
   |
note: multiple `impl`s satisfying `_: From<A>` found
  --> src/lib.rs:2:1
   |
 2 | impl From<A> for u32 {
   | ^^^^^^^^^^^^^^^^^^^^
...
 7 | impl From<A> for usize {
   | ^^^^^^^^^^^^^^^^^^^^^^
   = note: required for `A` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
   |
14 -     a.into() + 1
14 +     <A as Into<T>>::into(a) + 1
   |
```

So I think we should avoid introducing multiple `Into<{num}>` traits on the same type. Besides, the current code isn't significantly worse than `usize::from`/`Into::<usize>::into`, right?

arthurpaulino replied:
> **EDIT:** Oh, actually, the example won't work even if there is only one `Into<{num}>` impl. It really does not work well with type inference...

I tried a bit harder, but still...
```
error[E0284]: type annotations needed
 --> src/lib.rs:9:16
  |
9 |     1usize + a.into()
  |            -   ^^^^
  |            |
  |            type must be known at this point
  |
  = note: cannot satisfy `<usize as Add<_>>::Output == usize`
help: try using a fully qualified path to specify the expected types
  |
9 -     1usize + a.into()
9 +     1usize + <A as Into<T>>::into(a)
  |
```

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
